python/pyspark/sql/types.py

```diff
@@ -2256,11 +2256,33 @@ def convert(self, obj: datetime.timedelta, gateway_client: GatewayClient) -> Jav
         )
 
 
+class NumpyScalarConverter:
+    @property
```

Comment:
Shall we just a new variable and add a method in `pyspark.sql.utils`? Since this happens in the driver side only, I think we can just cache it in the module level.

Reply:
Sounds good! Defined `has_numpy` in `pyspark.sql.utils`.